### ISSUE TITLE
Kestrel failing on coreclr runtime beta7-12264 (Arch Linux)

### ISSUE THREAD

**Issue by ganeshran** [aspnet/Home#761]
I am trying to run the ASP.NET HelloMVC sample on ArchLinux. beta7-12264. It works fine on the mono runtime but gives filenotfound exception with coreclr.

The DLL which is not found is 'System.Runtime.Serialization.Primitives, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' 

```
>
error   : [Microsoft.AspNet.Diagnostics.ErrorPageMiddleware] An unhandled exception has occurred while executing the request
System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.Serialization.Primitives, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
>
File name: 'System.Runtime.Serialization.Primitives, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' ---> System.IO.FileNotFoundException: Could not load the specified file.
File name: 'System.Runtime.Serialization.Primitives'
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.Resolve(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
   at Microsoft.AspNet.Mvc.MvcJsonMvcOptionsSetup.<>c__DisplayClass0_0.<.ctor>b__0(MvcOptions _)
   at Microsoft.Framework.OptionsModel.OptionsManager`1.<>c__DisplayClass5_0.<Configure>b__1(TOptions options, IConfigureOptions`1 setup)
   at System.Linq.Enumerable.Aggregate[TSource,TAccumulate](IEnumerable`1 source, TAccumulate seed, Func`3 func)
   at Microsoft.Framework.OptionsModel.OptionsManager`1.GetNamedOptions(String name)
   at Microsoft.AspNet.Mvc.ApplicationModels.DefaultApplicationModelProvider..ctor(IOptions`1 mvcOptionsAccessor)
```

**Comment by Eilon** [comment]
@halter73

**Comment by Eilon** [comment]
Oh wait this doesn't really look like Kestrel at all...

@rynowak ?

**Comment by ganeshran** [comment]
@halter73  @rynowak 
kestrel executes without an error but when I open the web page http://localhost:5004, there is an internal server error. 

Below is the output and commands I used

>
ganesh@localhost:~/Development/dotnet/Home/samples/latest/HelloMvc|dev⚡
⇒  dnvm use 1.0.0-beta7-12264 -r coreclr
Adding /home/ganesh/.dnx/runtimes/dnx-coreclr-linux-x64.1.0.0-beta7-12264/bin to process PATH
ganesh@localhost:~/Development/dotnet/Home/samples/latest/HelloMvc|dev⚡
⇒  dnx . kestrel
Started
error   : [Microsoft.AspNet.Diagnostics.ErrorPageMiddleware] An unhandled exception has occurred while executing the request
System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.Serialization.Primitives, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
File name: 'System.Runtime.Serialization.Primitives, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' ---> System.IO.FileNotFoundException: Could not load the specified file.
File name: 'System.Runtime.Serialization.Primitives'

This is the page I see at localhost:5004 - http://i.imgur.com/x2Wfvhc.png

**Comment by halter73** [comment]
@ganeshran Thanks for your bug report. I use Arch too :smile:, but I've mainly been testing on mono.

I am able to reproduce the FileNotFoundException on my machine, but I haven't figured out the root cause yet. I'll try to take a look at it with @rynowak tomorrow.

**Comment by halter73** [comment]
@ganeshran It looks like you can resolve this issue by adding `"dependencies": { "System.Runtime.Serialization.Primitives": "4.0.10-*" }` to the `frameworks.dnxcore50` section of your project.json and restoring again.

The same FileNotFoundException gets thrown on the Windows coreclr when running `dnx . web` (WebListener). So maybe it's just an issue with the sample? Then again, I would expect `dnu build` to fail if a dependency is missing but it doesn't.

**Comment by ganeshran** [comment]
@halter73  Thank you. After adding the dependency, it works even on coreclr runtime :)

**Comment by ganeshran** [comment]
@halter73 What's the difference between dnu restore and dnu build?. I had run dnu restore but not build.

Are there any binaries created during the build process that are missing during restore?

**Comment by halter73** [comment]
@ganeshran `dnu restore` should the only command you need to retrieve your dependencies.

`dnu build` compiles your project and drops assemblies in your build folder (much like MSBuild's default Build target).

The only reason I brought up `dnu build` is because I was surprised HelloMvc would compile only to fail at runtime due to a missing dependency. Of course, this wouldn't be entirely unexpected if there was an `Assembly.Load` your something similar on the stack. But the exception was thrown from [MvcJsonMvcOptionsSetup](https://github.com/aspnet/Mvc/blob/48bfdceea6d243c5ec8d6e00f450f8fe7cce59f7/src/Microsoft.AspNet.Mvc.Formatters.Json/MvcJsonMvcOptionsSetup.cs), and I don't see anything like that there.

**Comment by ganeshran** [comment]
Hi @halter73 Yes I checked the  MvcJsonMVCOptionsSetup class too. It's calling its parent Options class's constructor and there too isn't any call to load the DLL dynamically.  

But your solution of explicitly adding the dependency works,  so should I close this issue or will it be treated as a bug since dnu restore should have detected this dependency itself?

**Comment by davidfowl** [comment]
It's actually a bug in JSON.NET because it isn't specifying any dependencies.

/cc @JamesNK

**Comment by ganeshran** [comment]
Thanks @davidfowl!

**Comment by JamesNK** [comment]
I'm not doing the what now?

The Json.NET assembly is a PCL and it isn't referencing anything external. The runtime It should figure out how to load stuff on its own IMO.

**Comment by davidfowl** [comment]
@JamesNK nope. It's a bug in the JSON.NET package. If you look at how new libraries are built, they specify their dependencies in the nuspec. Even the new PCL tooling expects this. It's the only way you can determine that a package actually works on some platform. Read this blog post http://oren.codes/2015/06/16/demystifying-pcls-net-core-dnx-and-uwp-redux/.

TL;DR Newtonsoft.Json needs the nuspec to have a dotnet folder and a dotnet dependency group that specifies all of the dependencies used to create that library.

**Comment by davidfowl** [comment]
Closing out this issue, I filed a bug here https://github.com/JamesNK/Newtonsoft.Json/issues/618
